internal/internal_workflow_client.go

```diff
@@ -1708,45 +1747,90 @@ func (w *workflowClientInterceptor) UpdateWorkflow(
 			},
 		},
 	})
-	handle := &updateHandle{ref: resp.GetUpdateRef()}
 	if err != nil {
-		handle.err = err
-		return handle, nil
+		return nil, err
 	}
 	switch v := resp.GetOutcome().GetValue().(type) {
 	case nil:
-		panic("unspported update outcome: Incomplete")
+		return &lazyUpdateHandle{
+			client:           w.client,
+			baseUpdateHandle: baseUpdateHandle{ref: resp.GetUpdateRef()},
+		}, nil
+	case *updatepb.Outcome_Failure:
+		return &completedUpdateHandle{
+			err:              w.client.failureConverter.FailureToError(v.Failure),
+			baseUpdateHandle: baseUpdateHandle{ref: resp.GetUpdateRef()},
+		}, nil
+	case *updatepb.Outcome_Success:
+		return &completedUpdateHandle{
+			value:            newEncodedValue(v.Success, w.client.dataConverter),
+			baseUpdateHandle: baseUpdateHandle{ref: resp.GetUpdateRef()},
+		}, nil
+	}
+	return nil, fmt.Errorf("unsupported outcome type %T", resp.GetOutcome().GetValue())
+}
+
+func (w *workflowClientInterceptor) PollWorkflowUpdate(
+	ctx context.Context,
+	in *ClientPollWorkflowUpdateInput,
+) (converter.EncodedValue, error) {
+	// header, _ = headerPropagated(ctx, w.client.contextPropagators)
+	//todo header not in PollWorkflowUpdate
+
+	grpcCtx, cancel := newGRPCContext(ctx, defaultGrpcRetryParameters(ctx))
+	defer cancel()
+	wait := updatepb.WaitPolicy{
+		LifecycleStage: enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED,
+	}
+	pollReq := workflowservice.PollWorkflowExecutionUpdateRequest{
+		Namespace:  w.client.namespace,
+		UpdateRef:  in.UpdateRef,
+		WaitPolicy: &wait,
+		Identity:   w.client.identity,
+	}
+	resp, err := w.client.workflowService.PollWorkflowExecutionUpdate(grpcCtx, &pollReq)
```

Comment:
This needs to be in a loop and the server needs to wait no longer than 60s before sending back a "not yet, try again". This helps with our philosophy of promising to proxy creators and others that we have no call lasting longer than one minute.

Reply:
Server is just going to return a GRPC timeout status here, it won't (and IMO shouldn't) do the early-return trick. But I can detect that and retry if the parent context still has some life left in it.

Reply:
I think we should do the early return "trick". I don't think it's a failing state to have a poll say "nothing yet". This is different than a timeout which the caller cannot differentiate between a _real_ timeout (client or server side based on context deadline) and this one.

Whatever we choose to do, I think it needs to never take longer than a worker poll. And I believe we are telling people with proxies to set their max RPC timeouts at 80s. And the client should retry forever (or as long as their caller is willing) in the scenario where it isn't done yet. I think when you write the if statement to have the client retry on gRPC code timeout, it'll become clear that reusing timeout here is not the best.

Reply:
I would say since the retrying is hidden from the user, whether or not the early-return thing is used isn't relevant except to make it easy for us to detect and retry (and possibly the proxy thing, but I don't know if we actually advise users that).

The early return does feel slightly lame, though I can't really pinpoint exactly why. And actual timeout feels more semantically appropriate to me, and we can distinguish which one is which by including a bit of metadata.

As much as it's not my favorite argument, I can kinda get sticking with the early return to keep things consistent w/ other long pollers

Reply:
Fixed up to loop until the parent context .Err() returns non-nil. The server is definitely returning a GRPC timeout status right now (and that was an intentional design choice) but I can verify that we're happy with the approach again.

Reply:
Yes, I am basically trying to make sure these expected timeouts won't count against our SLAs

Reply:
Yep, confirmed with the server team who requested this style in the first place.